### ISSUE TITLE
Should we allow requests without timeout?

### ISSUE THREAD

**Issue by Neverlord** [actor-framework/actor-framework#442]
This code is currently (taking about the 0.15 topic branch) legal and even idiomatic in CAF:

```cpp
  // send "Hello World!" to our buddy ...
  self->request(buddy, "Hello World!").then(
    // ... wait for a response ...
    [=](const string& what) {
      // ... and print it
      aout(self) << what << endl;
    }
  );
```

This usually works fine. However, `request` is asynchronous message passing at the end of the day. If `buddy` is overloaded, a response can take any amount of time. In a distributed setting, the message even could get lost with our current BASP implementation. Say node `A` is sending a message to `D` but has no direct connection to it. It is connected to `B` and `C`, each offering a forwarding path to `D`. If `A` sends it message via `B` and `B` fails, the message is lost, event though `A` can still reach `D` via `C`.

It's another discussion whether it is worth it to make sure message delivery is guaranteed (a question which @josephnoir hopefully answers in the next couple of weeks). In case all nodes are connected directly, `D` could still fail. In 0.14, we silently kept track of in-flight requests and would generate an error message if the destination of a request died. This implicitly adds monitoring overhead for each actor we send a request to. Is this a "fair price" to pay? Is it desirable?

A straightforward solution to this problem would be to make timeouts mandatory. This could either use the `after(...) >> callback` syntax, or `request` always take a timeout specification as second argument:

```cpp
  // send "Hello World!" to our buddy ...
  self->request(buddy, seconds(30), "Hello World!").then(
    // ... wait up to 30 seconds for a response ...
    [=](const string& what) {
      // ... and print it
      aout(self) << what << endl;
    }
  );
```

In the snippet above, we don't provide a handler for error messages. This means the actor dies if its request results in an error (such as the 30s timeout). In this way, users write minimal code, but we can enforce timing restriction throughout the system. This could of course be combined with our current "silent monitoring" if desirable to have a "shortcut" when sending to dead actors.

Opinions?

**Comment by mavam** [comment]
> In a distributed setting, the message even could get lost with our current BASP implementation.

Is this effectively creating a memory leak?

>  In 0.14, we silently kept track of in-flight requests and would generate an error message if the destination of a request died. This implicitly adds monitoring overhead for each actor we send a request to. Is this a "fair price" to pay? Is it desirable?

Can you give us a better sense of the performance penalty where speaking about? Is it linear in the number of outstanding requests? Have You Measured It?

> A straightforward solution to this problem would be to make timeouts mandatory.

There's a camp of Erlang folks out there who religiously believe in timeouts, and that programmers always should annotate their code with upper latency bounds. This also helps reading code and understanding the intended architecture better. At the same time, it can be tedious in cases when latency expectations are uniform.

Assuming the performance penalty turns out to be significant, I could image the following: each actor system has a default timeout that users can control. In addition, users can override this timeout per request where they see fit.

**Comment by Neverlord** [comment]
> Is this effectively creating a memory leak?

Not a leak, but a possible deadlock on the sender who is waiting indefinitely for a response.

> Can you give us a better sense of the performance penalty where speaking about?

It's not about performance. More about extra state and thus higher memory consumption per actor.

> I could image the following: each actor system has a default timeout that users can control. In addition, users can override this timeout per request where they see fit.

I don't like hidden defaults. If I'd have a timeout I consider to cover 99% of all cases, I would prefer it to have single configuration header with constants for such values per project. It makes it clear what is happening is the code.

Also, I tend to side with the Erlang camp, but not for religious reasons. ;)

**Comment by Neverlord** [comment]
>  I still don't have a clear picture of a concrete use case where the memory consumption becomes an issue

To clarify this a bit: every single Byte we put into each actor by adding members to `local_actor` multiplies to the thousands for each CAF application. I always ask "what could we possibly take away?" and "do we really need this?"

> By using requests, users know that a response may never arrive and can get lost.

Do you really think this assumption is correct? Think about how to teach CAF to new users and how steep their learning curve is. Hidden information is always bad. If you look at the call side and see a time parameter, you can just look up what this parameter is doing and you know what is happening. If your actor dies because it receives a timeout that you have never specified, would you think this to be good API?

And how would you set this default timeout in the first place? One second? Ten? 30 Minutes? CAF is a general purpose framework, so we can't make assumptions on user-written applications. A backend for a web site has different time requirements than Big Data applications or high-speed trading tools.

Let me rephrase my question and add a bit more background.

When using good ol' `send` and `become`, our example would look like this:

```cpp
  // send "Hello World!" to our buddy ...
  self->send(buddy, "Hello World!");
  self->become(
    // ... wait up to 30 seconds for a response ...
    [=](const string& what) {
      // ... and print it
      aout(self) << what << endl;
    },
    after(seconds(30)) >> [] {
      // ... ??? ...
    }
  );
```

Of course, `after(...)` is optional. So the question is: do we design `request` in the same way or do we make the timeout mandatory instead? The point is: CAF does not know what you are doing when calling `become`. It most likely is your default behavior and you are waiting for some work. However, when using `request`, we already know exactly that the handler you give it handles the response message. And that response might never arrive. So, this is a potential deadlock by definition.

Enforcing the time parameter to `request` also streamlines the API. You can already omit the `error` handler and your actor just dies per default. Turning timeouts into `error` messages as well saves typing and IMO keeps the code more readable.

So this is the worst case we are talking about:

```cpp
 self->request(buddy, "Hello World!").then(
    [=](const string& what) {
      // ... handle result ...
    },
    [=](const error& err) {
      // ... handle error ...
    },
    after(seconds(30)) >> [=] {
      // ... handle timeout ...
    }
  );
```

With a required time parameter, users still can distinct between "actual errors" and timeouts, but no longer *have* to:

```cpp
 self->request(buddy, seconds(30), "Hello World!").then(
    [=](const string& what) {
      // ... handle result ...
    },
    [=](const error& err) {
      if (err == sec::request_timeout) {
        // ... handle timeout ...
      } else {
        // ... handle error ...
      }
    }
  );
```

As long as users are fine with just killing the actor on either errors or timeouts, they can omit the error handler completely:

```cpp
 self->request(buddy, seconds(30), "Hello World!").then(
    [=](const string& what) {
      // ... handle result ...
    }
    // die when receiving an error or timeout
  );
```

Talking about how to teach CAF: if we go with the optional-`after(...)`-route, our advice for new users basically would be: never use requests without timeouts for reasons X, Y, and Z. This is why I started this discussion: isn't it better to just hard-wire this into the API? This also gives us the opportunity to have a more streamlined API and less code as shown above.

Sorry if my initial post didn't make those points clear.

**Comment by josephnoir** [comment]
Let's see if I got this: In 0.14 sending a request automatically leads to an implicit monitor to notify the actor if it's communication partner becomes unreachable. However, messages can get lost as described in your scenario above. As a result, we cannot guarantee that the message is either delivered or that an error is received and that's why a deadlock can happen.

First of all, what is gained by implicit monitoring the destination node, is the implication of an request that either a result or an error is received? Yes, you have already answered that here:

> However, when using request, we already know exactly that the handler you give it handles the response message. And that response might never arrive. So, this is a potential deadlock by definition.

I'd argue that the implicit monitor isn't helpful as it does not solve the problem, but only soothes you (I am missing a better word). This bothers me more than a potential memory overhead.

Anyway, the alternative we are discussing is making a timeout mandatory. I think it is a good thing if the API reflects the best practice. There are two things I am wondering about:

1. You said *actual errors and timeouts* which implies that timeouts are a different category of errors? Is that the case and if so, do we want to mix those here?

2. Are there scenarios where you don't want to specify a timeout, e.g., in a local scenario and how does this affect updating existing code--is it straight forward or do we break scenarios?

Another "fix" would be to guarantee the delivery of a request if no link / monitor error is received. I am sure that implementation is much more complex, but would it also be more desirable from a user perspective?

**Comment by Neverlord** [comment]
> Let's see if I got this: In 0.14 sending a request automatically leads to an implicit monitor to notify the actor if it's communication partner becomes unreachable.

Yes. Which isn't without problems as we have already discovered here. :)

> I'd argue that the implicit monitor isn't helpful as it does not solve the problem

Correct.

If we have mandatory timeouts, the only benefit of implicit monitoring is to not wait until the timeout occurs if we already know the receiver is dead.

In a world without mandatory timeouts (i.e. the current implementation), this at least prevents *some* deadlocks (which are otherwise really easy to run into).

> You said actual errors and timeouts which implies that timeouts are a different category of errors?

Whether or not timeouts are are treated differently is up to the user. If you look at my last three code examples: if we go with a separate `after()...` handler, users *have* to treat them differently. If we turn timeouts into `error` messages, users *can* treat them differently, but don't have to and get the kill-on-error default by omitting explicit error/timeout handling.

> how does this affect updating existing code

0.15 is a breaking API change anyways, so that's not a concern in my opinion. It's more important to think about what is the right thing to do going forward.

> Another "fix" would be to guarantee the delivery of a request if no link / monitor error is received. I am sure that implementation is much more complex, but would it also be more desirable from a user perspective?

Sure, guaranteed delivery would be nice. But the runtime systems needs to be scalable. Putting state into the runtime system can blow up really quick. You also have to make sure no request is duplicated along the way.

A simple example. An actor on node `A` sends a request to another actor on node `D`. A is not directly connected to `D` but can forward via `B` and `C`. If `B` forwards a request, it has to signalize `A` back once it forwarded a request. Because, if `B` fails before `A` knows `C` got the message, it has to re-send the request using a different path. But what if `B` fails after it forwarded the message to `C` but before it could report back to `A`? So, we need to detect duplicated requests at `C` anyway if `A` resends a request that it falsely assumed didn't arrive. You have to do this for each request in the distributed system.

That was the easy part. What about forwarding or delegating requests? If the request is delegated from `C` to `E`, we have a responsibility chain, where `A` has state it needs for the request to `C` and `C` has state to avoid duplication and also has state to transmit the delegated request to `E`. What's worse: even if `E` has a direct connection to `A`, we still have to signalize success back to `C` to get rid of the state.

This is also a problem with relying on monitoring, btw. If we monitor the actor that receives the initial request, we have no idea whether or not it did forward or delegate this request before dying. Timeouts do have their own problems, but at least they enable local decision making and don't cause exploding state in the runtime system.

Long story short: I do not think guaranteed delivery scales. Even in a small cluster with "only" a few dozen nodes, each connected directly: keeping track of each and every request that crosses the wire? I would be happy if you prove me wrong, though. ;)

**Comment by mavam** [comment]
> If your actor dies because it receives a timeout that you have never specified, would you think this to be good API?

It may. In this particular case the user would just look up the actor system timeout, which either the user or the developer themselves specified at some point. 

> And how would you set this default timeout in the first place? One second? Ten? 30 Minutes? CAF is a general purpose framework, so we can't make assumptions on user-written applications. A backend for a web site has different time requirements than Big Data applications or high-speed trading tools.

That is exactly up to the developer who wrote the application. CAF certainly cannot provide a one-size-fits-all default. What I've been advocating so far is that CAF would not assume a timeout by default.

I like your code examples, and especially the differentiated way to treat timeouts. The part I am not convinced about is whether

```cpp
self->request(buddy, seconds(30), "Hello World!").then(
```

should be mandatory and cannot be written as

```cpp
self->request(buddy, "Hello World!").then(
```

This boils down to the original question: is it too much of a burden to specify a timeout? What if the programmer cannot make a good choice at this point? Do we want to force them and expect it to fix later when actors terminate at runtime? I fully agree that this is easier to debug than a deadlock. If a developer cannot decide at this point what they should put in there, what are they gonna do if they are afraid of choosing a destructive upper bound? Write `request(x, minutes(10), ...)`?

To be clear: I'm not strictly against mandatory timeouts, but I'm not yet 100% convinced that this change doesn't come without significant usability concerns. I wonder whether it's possible to give users the choice of enforced timeouts vs. optional specification. Could this be part of an actor system configuration? With mandatory timeouts enabled, an actor would terminate when attempting to send a request without timeout. When disabled, the request would go through just fine. (Yes, compile-time would nicer...)

**Comment by Neverlord** [comment]
> If a developer cannot decide at this point what they should put in there, what are they gonna do if they are afraid of choosing a destructive upper bound? Write `request(x, minutes(10), ...)`?

Honestly, I think 10 minutes isn't a bad start. As long as the system eventually reacts to an unexpected behavior and either falls back to a defined recovery state or at least crashes as a whole, it's fine.

> To be clear: I'm not strictly against mandatory timeouts, but I'm not yet 100% convinced that this change doesn't come without significant usability concerns.

I wouldn't have started the discussion if didn't share your concerns. Still, safety should come first. I have strong reservations towards making a runtime parameter that has such a huge impact on the system. Correctness should be an inherent feature of a framework like CAF and not something users can turn off at will or---even worse---by accident.

Regarding overhead: I think a good template for CAF applications would look somewhat like this:

```
/project-root
|-- includes
|   `-- typed_actors.hpp
|   `-- atoms.hpp
|   `-- timeouts.hpp
|-- src
|   `-- main.cpp
|-- CMakeLists.txt
```

With `typed_actors.hpp` containing all typedefs for `typed_actor<...>`, `atoms.hpp` containing all `atom_constant<...>` aliases, and `timeouts.hpp` containing the maximum durations for different operations. Offering such a template as Git repository and linking it in the "Getting Started" section in the manual could give programmers a head start.

**Comment by mavam** [comment]
> Honestly, I think 10 minutes isn't a bad start. As long as the system eventually reacts to an unexpected behavior and either falls back to a defined recovery state or at least crashes as a whole, it's fine.

I recovering system (even if it means crash) would be ideal. But isn't that application-dependent? What if 15 minutes would have been better for this deployment? The problem with timeouts at points where they are a longshot is that they are hard to refactor and calibrate. I haven't had that experience yet, but I fear that hard-coding timeouts at places where there is no good default value drives up the development cost.

> Still, safety should come first. I have strong reservations towards making a runtime parameter that has such a huge impact on the system.

Yes, safety first. In this light, a default timeout for an entire actor system would offer too much leverage. I'm no longer a proponent. 

>  [..] and `timeouts.hpp` containing the maximum durations for different operations.

Would this lead to code like `request(x, default_timeout, ...)`? I do not see a significant difference to an actor-system-wide default, with the exception that you now have to coordinate even more default values.

At this point I'm leaning against mandatory timeouts, but also against default timeouts. I would still advocate that developers should have the choice. Those who build large-scale systems should understand that timeouts represent an inherent safety net, in addition to communicating latency expectations to other readers of the code. In my eyes, it should be "good style," but not mandatory because CAF does not solely focus on low-latency and high-throughput use cases.

**Comment by Neverlord** [comment]
> it should be "good style," but not mandatory because CAF does not solely focus on low-latency and high-throughput use cases.

I don't think this is related to low-latency or high-throughput applications. It is fundamental decision regarding robustness to failure.

The question is: what is the alternative? Requests without timeouts can cause deadlocks even in non-distributed scenarios. If the API offers a request without timeouts, than this could give developers a false sense of security. The primary purpose of timeouts is not to maintain strict timing requirements but to guarantee liveness and to make sure errors are detected eventually. Always.

Programmers can already shoot themselves in the foot by using good ol' `send` and wait for a response by using `become`. By adding `request` to the mix, we have the opportunity to not only give developers a more idiomatic, convenient and concise interface, but also to make sure basic safety mechanisms are in place.

Version 0.15 embodies our vision for a CAF 1.0 release. It will break a lot of code, but should stabilizes the API going forward. With a broadening user base, we can't afford to break the API every few releases. We can still add features to the API later, but not easily remove it.

So, my proposal is at follows. We 0.15 release only contains a `request` *with* timeout. This sacrifices some convenience for the sake of safety and reliability.

**Comment by josephnoir** [comment]
My comment on reliable delivery was in no way meant to imply that it is the preferable solution. Simply wanted the idea to be include it in the discussion and have us think about it before discarding the idea. The state it brings is one of the biggest problems here. The state we have for timers is probably much easier to keep as it is only local, but how complicated is it to track timeouts for more than a million messages?

> To be clear: I'm not strictly against mandatory timeouts, but I'm not yet 100% convinced that this change doesn't come without significant usability concerns.

Isn’t running into deadlocks a significant usability concern? 

> Requests without timeouts can cause deadlocks even in non-distributed scenarios.

If I am not mistakes your earlier examples include forwarding via a third node, how does this work without distribution? (Simply by sending a message, which may never be processed?)

None the less, I can see that setting the timeout in every request call seems like a hassle. I’d be interested to get isight from more developers how are using CAF to build bigger projects. For now, we have @mavam who does not seem to be convinced! (yet)

How complicated is it to change the syntax in the `actor_system` branch? We could see how people already using the branch react to it … iirc there were a few people on the mailing list who do that.

btw: Considering one might want to resend a message when running into a timeout. Is there a nice way to do that?

**Comment by mavam** [comment]
I'm going to comment on @josephnoir's part first.

>  The state we have for timers is probably much easier to keep as it is only local, but how complicated is it to track timeouts for more than a million messages?

The overhead of timeouts is an interesting related question. Are there any foreseeable scalability concerns? (I do not know how the runtime keeps track of timers, but would assume that a dedicated timer manager creates a total order of expiration time points.)

> Isn’t running into deadlocks a significant usability concern?

Sure, you can see it that way. At the end of the day almost everything is a usability concern :-). To be more precise, I mean usability in the sense of API design. A developer who is in rapid prototyping mode and doesn't want to think about timeout when creating the basic communication structure might just carelessly put in large values to bypass the annoyance. This renders timeouts effectively useless in practice. That is my chief concern and the tension point of this discussion.

> How complicated is it to change the syntax in the actor_system branch? We could see how people already using the branch react to it … iirc there were a few people on the mailing list who do that.

Changing the API is really just a matter of adjusting the signature of `request`. Sending a poll to the mailing list could be interesting data.

> btw: Considering one might want to resend a message when running into a timeout. Is there a nice way to do that?

I asked that question in the very first days of libcppa. I'm sure we'll get it more often in the future and I advocate to include it in the FAQ. For now, CAF does not provide machinery for resending timed-out messages. It's actually a delicate topic, since it introduces buffering on the sender side. How should CAF provision the right amount of buffer? What happens when the buffer is full? In a single-machine scenario, what is happening to the reference count of the message at the receiver side (CAF should would have to ensure that the sender releases the message before the receiver dispatches it to allow ownership transfers)? There is no one-size-fits-all answer to these problems.

To @Neverlord's points:

> Version 0.15 embodies our vision for a CAF 1.0 release. It will break a lot of code, but should stabilizes the API going forward. With a broadening user base, we can't afford to break the API every few releases. We can still add features to the API later, but not easily remove it.

No doubt, I fully agree.

> So, my proposal is at follows. We release 0.15 with a request function that requires a timeout. This sacrifices some convenience for the sake of safety and reliability. If we later figure out a good way how to propagate errors even without timeouts (without putting too much pressure on the runtime system), we can still add this function later on.

I would be fine with that if you convince me that we're not losing the rapid prototypers, who will punch in a timeout several orders of magnitude above what they're expecting to get around the API and get something that works, with the idea to later tighten down the timeouts in a second pass.

**Comment by Neverlord** [comment]
> I’d be interested to get isight from more developers how are using CAF to build bigger projects.

Sure, but that wouldn't solve this discussion either. If they simply aren't aware that their design is flawed and can run into deadlocks, then what have you learned? It's not a reasonable thing to advise people to write unsafe code because everyone is doing it.

> > Requests without timeouts can cause deadlocks even in non-distributed scenarios.

> If I am not mistakes your earlier examples include forwarding via a third node, how does this work without distribution? (Simply by sending a message, which may never be processed?)

Yes. When using dynamically typed actors, this is quite easy to run into.

> > btw: Considering one might want to resend a message when running into a timeout. Is there a nice way to do that?

> I asked that question in the very first days of libcppa. I'm sure we'll get it more often in the future and I advocate to include it in the FAQ.

Noted.

> The overhead of timeouts is an interesting related question. Are there any foreseeable scalability concerns?

There is a central timer (actor) that keeps track of them. Signalizing is of course done by messages, so you have one message to the timer, one `O(log n)` insert on a `std::multimap`, and one message back. The map is only accessed at the top for read / remove, so these operations are all `O(1)`.

If the central timer ever turns out to be a bottleneck, we could have several of those and load-balance by having actors access a timer dependent on the scheduler worker executing them or something along those lines. I don't expect this to be a problem, though.

> I would be fine with that if you convince me that we're not losing the rapid prototypers, who will punch in a timeout several orders of magnitude above what they're expecting to get around the API and get something that works, with the idea to later tighten down the timeouts in a second pass.

Isn't an unrealistic timeout still better than no timeout? ;)

I'll go ahead and make timeouts mandatory for now but leave this issue open for discussion.

**Comment by mavam** [comment]
> Isn't an unrealistic timeout still better than no timeout? ;)

I don't think so. Now you have to have know *if* a timeout is off. Maybe it's easy to eyeball, but maybe not. Unless the code is self-descriptive or has a comment, good luck. Then you have to improve the timeout iteratively, and converge to an appropriate one during testing.

**Comment by Neverlord** [comment]
I think with the current solution with a required timeout that can be set to `infinite` is a good compromise.
